src/workerd/api/sockets.c++

```diff
@@ -119,4 +100,19 @@ jsg::Promise<void> Socket::close(jsg::Lock& js) {
   }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
 }
 
+jsg::Promise<void> Socket::handleProxyStatus(
+    jsg::Lock& js, kj::Promise<kj::HttpClient::ConnectRequest::Status> status) {
+  auto& context = IoContext::current();
+  return context.awaitIo(js, kj::mv(status),
+      [this](jsg::Lock& js, kj::HttpClient::ConnectRequest::Status&& status) -> void {
+    if (status.statusCode < 200 || status.statusCode >= 300) {
+      // If the status indicates an unsucessful connection we need to reject the `closeFulfiller`
+      // with an exception. This will reject the socket's `closed` promise.
+      auto exc = kj::Exception(kj::Exception::Type::FAILED, __FILE__, __LINE__,
+        kj::str(JSG_EXCEPTION(Error) ": proxy request failed"));
+      resolveFulfiller(js, exc);
+    }
+  });
```

Comment:
While it's unlike this will typically fail in the common case, it's probably worthwhile marking the returned JS promise here as handled so that if it does fail it will not trigger an unhandledrejection event. Alternatively, an exception handler can be added to log any rare exception that may occur here.